### PR TITLE
feat(controller): Add HTTP template

### PR DESCRIPTION
**Pull request by alexec**

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 	"time"
 
 	"github.com/argoproj/argo-workflows/v3/util/env"
+	"github.com/argoproj/argo-workflows/v3/workflow/controller/httptemplate"
```
Comment (alexec) [inline workflow/controller/controller.go:12]: @simster7 I've added this, not to show how to do HTTP template, but how I think we should try and do templates in general. 

I've located all code for this template in another package and broken the coupling between `woc` and the template. This keeps the code MUCH cleaner. The will also be possible for `PodTemplate`, so I hope we can also do it for `DataTemplate`.

Comment (simster7) [reply workflow/controller/controller.go:12]: Will look at this when working on HTTP/Data templates

Comment (simster7) [reply workflow/controller/controller.go:12]: @alexec I like the segmentation between woc and template, but not sure if I'm a fan of having each template type be its own mini controller... I think breaking up the operator's single-threaded operation could introduce many bugs. Could you perhaps explain more of your reasoning during a standup?

Comment (alexec) [reply workflow/controller/controller.go:12]: For HTTP template, we don't want to be doing the requests inline as they could take a >30s. Instead, they have to be done async.That means we have to have a controller. That's actually the same case as with the container and script template. We make an HTTP request to the Kubernetes API and we then wait for the response. 

I'm not sure what you mean by "introduce bugs"? Please explain more?

Comment (simster7) [reply workflow/controller/controller.go:12]: > I'm not sure what you mean by "introduce bugs"? Please explain more?

The core idea of how the operator works right now is that it's processed in a single, uni-threaded, idempotent operation. When we allow for different threads to plug into it, we lose the innate safety provided by the idempotency. Possible issues that come up to mind: different results if wf is immediately requeued, concurrency issues, race conditions, etc.

Comment (alexec) [reply workflow/controller/controller.go:12]: Ah. Yes, you must write templates with all of those in mind. That isn't to do with the design, that is to do with the implementation.

**Comment by agilgur5** [comment]
Seems like this was replaced by #5750

**Pull request closed without merging**